### PR TITLE
Add a reusable queue system to use and abuse

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/70889

This PR adds a `Tribe_Queue` class to put in place a queueing system we can reuse:

- it's made to try to run anytime it can from `admin_head` to `wp_ajax_` events
- it will run on the `tribe_queue_work` action
- said action is scheduled each 30"

The use boils down to this:

```php
// file My_Object.php

class My_Object() {
    public function set_post_title( $id, $index, $data ){
        $title = $data[$index];

        $updated = wp_update_post( array( 'ID' => $id, 'post_title' => $title ) );

        return !empty( $updated );
    }
}

// elsewhere in code
tribe_singleton('my_object', 'My_Object');
$queue = tribe( 'queue' );

$targets = array( 2, 3, 4, 5, 6 );

// to avoid using (untestable) static methods (due to cron) the q will use the container to build stuff
// in plain English: "build `my_object` and call `set_post_title` on each target"
$callback = array( 'tribe', 'my_object', 'set_post_title');

// this can be anything
$data = array( 'foo', 'baz', 'bar', 'foo', 'baz' );

$queue->queue_work( $targets, $callback, $data)
    ->set_batch_size(3) // on how many targets per run?
    ->set_priority( 5 ) // same as WP priority system, lowest is first
    ->save(); // actually commit to queue this
```

If the callback function returns anyting "truthy" then the queue will deem the work on the single target done else it will move the target at the end of  the queue and try again later.

### CODE REVIEW

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | The method parameter $set_b is never used | [Average_Proximity_Start_Detector.php:37](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Average_Proximity_Start_Detector.php#L37)
:warning: | The method parameter $set_a is never used | [Average_Proximity_Start_Detector.php:81](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Average_Proximity_Start_Detector.php#L81)
:warning: | The method parameter $set_b is never used | [Average_Proximity_Start_Detector.php:81](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Average_Proximity_Start_Detector.php#L81)
:warning: | The method parameter $b_set is never used | [Closest_Unique_Start_Detector.php:56](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Closest_Unique_Start_Detector.php#L56)
:warning: | The method parameter $set_b is never used | [Closest_Unique_Start_Detector.php:94](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Closest_Unique_Start_Detector.php#L94)
:warning: | The method parameter $set_b is never used | [Closest_Unique_Start_Detector.php:206](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Closest_Unique_Start_Detector.php#L206)
:warning: | The method parameter $segment is never used | [Closest_Unique_Start_Detector.php:224](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Closest_Unique_Start_Detector.php#L224)
:warning: | The method parameter $b_ends is never used | [Closest_Unique_Start_Detector.php:224](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Closest_Unique_Start_Detector.php#L224)
:warning: | The method parameter $report is never used | [Closest_Unique_Start_Detector.php:224](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Closest_Unique_Start_Detector.php#L224)
:warning: | The method parameter $set_b is never used | [Detection_Strategy.php:77](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Detection_Strategy.php#L77)
:warning: | The method parameter $set_b is never used | [Detection_Strategy.php:183](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Detection_Strategy.php#L183)
:warning: | The method parameter $b_set is never used | [Detection_Strategy.php:249](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Detection_Strategy.php#L249)
:x: | empty() and isset() must only contain variables (PHP 5.2 compatibility) | [Median_Proximity_Start_Detector.php:19](https://github.com/moderntribe/tribe-common/blob/1a27b1eb7ce91007d82cc34d7926649f6781da51/src/Tribe/Collisions/Median_Proximity_Start_Detector.php#L19)

> via jenkins/codesniffer-phpcs

File: src/Tribe/Cron.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+	/**
+	 * Filters the cron schedules to add intervals we need.
+	 *
+	 * @param array $schedules
+	 *
+	 * @return array
+	 */
+	public function filter_cron_schedules( array $schedules ) {
+		$schedules[ 'thirty_seconds' ] = array(
+			'interval' => 30,
```
Comment (bordoni) [inline src/Tribe/Cron.php]: As a question, you know that by default WordPress won't allow us to do that right? It only allows us to do I think (dont quote me on that) 15 minutes, but we can _ "force"_ it.

Comment (lucatume) [inline src/Tribe/Cron.php]: Older versions will not allow for that, but newer versions will. 
In a service project we have a 20" cron.

Comment (lucatume) [inline src/Tribe/Cron.php]: We try at least.

**Comment by lucatume** [comment]
Ok, I've moved away from `CONST` and put in place a priority system

File: src/Tribe/Cron.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+	/**
+	 * Filters the cron schedules to add intervals we need.
+	 *
+	 * @param array $schedules
+	 *
+	 * @return array
+	 */
+	public function filter_cron_schedules( array $schedules ) {
+		$schedules[ 'thirty_seconds' ] = array(
```
Comment (bordoni) [inline src/Tribe/Cron.php]: Remove spaces around `''`
